stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -616,166 +335,73 @@ private void releaseChildNodeRemovedEvent(ChildNodeRemovedEvent childNodeRemoved
+            element);
+      } else {
+        InspectNodeRequestedEvent message = new InspectNodeRequestedEvent();
+        message.nodeId = nodeId;
+        mPeerManager.sendNotificationToPeers("DOM.inspectNodeRequested", message);
+      }
     }
 
-    public void removeWithEvent(int index) {
-      Object element = remove(index);
+    public void onChildNodeRemoved(Object element) {
       ChildNodeRemovedEvent removedEvent = acquireChildNodeRemovedEvent();
-      removedEvent.parentNodeId = mParentNodeId;
-      removedEvent.nodeId = mObjectIdMapper.getIdForObject(element);
+
+      ElementInfo oldElementInfo = mDocument.getShadowDOM().getElementInfo(element);
+      removedEvent.parentNodeId = mDocument.getObjectIdMapper().getIdForObject(oldElementInfo.parentElement);
+      removedEvent.nodeId = mDocument.getObjectIdMapper().getIdForObject(element);
       mPeerManager.sendNotificationToPeers("DOM.childNodeRemoved", removedEvent);
+
       releaseChildNodeRemovedEvent(removedEvent);
     }
-  }
 
-  private final class DOMObjectIdMapper extends ObjectIdMapper {
-    @Override
-    protected void onMapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+    public void onChildNodeInserted(Object element, int parentNodeId, int previousNodeId, DOMView view, Accumulator<Object> insertedElements) {
+      ChildNodeInsertedEvent insertedEvent = acquireChildNodeInsertedEvent();
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.hook(object);
-    }
+      insertedEvent.parentNodeId = parentNodeId;
+      insertedEvent.previousNodeId = previousNodeId;
+      insertedEvent.node = createNodeForElement(view, element);
 
-    @Override
-    protected void onUnmapped(Object object, int id) {
-      mDOMProvider.verifyThreadAccess();
+      insertedElements.store(element);
+      mPeerManager.sendNotificationToPeers("DOM.childNodeInserted", insertedEvent);
 
-      NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(object);
-      descriptor.unhook(object);
+      releaseChildNodeInsertedEvent(insertedEvent);
     }
   }
 
-  private final class ProviderListener implements DOMProvider.Listener {
-    @Override
-    public void onPossiblyChanged() {
-      updateTree();
-    }
-
+  private final class PeerManagerListener extends PeersRegisteredListener {
     @Override
-    public void onAttributeModified(Object element, String name, String value) {
-      mDOMProvider.verifyThreadAccess();
+    protected synchronized void onFirstPeerRegistered() {
+      // We don't call mDOMProvider.setListener() until the first DOM.getDocument request.
 
-      AttributeModifiedEvent message = new AttributeModifiedEvent();
-      message.nodeId = mObjectIdMapper.getIdForObject(element);
-      message.name = name;
-      message.value = value;
-      mPeerManager.sendNotificationToPeers("DOM.attributeModified", message);
+      mDocument.createDomProviderIfNoneExists();
```

Comment:
Prior to the refactoring, `DOM` created an instance of `DOMProvider` inside `OnFirstPeerRegistered`, and removed it in `OnLastPeerUnregistered`. `DOM` and `CSS` have unique peer managers, and `Document` does not/should not have a peer manager (since it is not a domain class). 

Now, since `Document` owns `DOMProvider`, but has no peer manager, it has to delegate the responsibility of creating and destroying DOMProvider to the domain classes (DOM, CSS, etc.). 

As of this moment, the only class that makes use of `Document` is `DOM`, which is why this logic remained in `DOM`. 

A possible solution to this would be to use reference counting, and take care of creating/destroying `DOMProvider` inside of `Document`.

Another solution is to just create a `DOMProvider` when the app starts, without worrying about whether or not there are peers.